lib/chef/node/immutable_collections.rb

```diff
@@ -33,16 +33,21 @@ def safe_dup(e)
       end
 
       def convert_value(value)
-        # The order in this case statement is *important*.
-        # ImmutableMash and ImmutableArray should be tested first,
-        # as this saves unnecessary creation of intermediate objects
         case value
-        when ImmutableMash, ImmutableArray
-          value
         when Hash
-          ImmutableMash.new(value, __root__, __node__, __precedence__)
+          if ImmutableMash === value
```

Comment:
In this case, every `Immutable{Mash,Array}` will get two class comparisons, one for `{Hash,Array}` and then another for `Immutable{Mash,Array}`.

The case where we aren't yet immutable has less comparisons (3 or 4 -> 2), but the case where we are immutable, which should be the majority, will now increase from 1 or 2 to always 2. And the fact that Mashes are far more common than arrays in the node means that we can basically treat it as 1 -> 2 (Since ImmutableHash is first), so we double the common path.

I'm not quite sure this is better.

Reply:
Here's a list of what can potentially hit Immutablize#convert_value, along with the number of comparisons,

| Class          | Before | After | Delta |
|----------------|--------|-------|-------|
| Hash           |      3 |     2 |    -1 |
| Mash           |      3 |     2 |    -1 |
| VividMash      |      3 |     2 |    -1 |
| ImmutableMash  |      1 |     2 |    +1 |
| Array          |      4 |     3 |    -1 |
| AttrArray      |      4 |     3 |    -1 |
| ImmutableArray |      2 |     3 |    +1 |
| Object         |      4 |     2 |    -2 |

I also made a benchmark script here: https://gist.github.com/dafyddcrosby/349dbfb86c2eec157060aa0685b25bbf